core/app/models/spree/promotion/legacy_adjuster.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module Spree
+  class Promotion < Spree::Base
+
+    # This class encapsulates all the things the promotion system does to
+    # an order.
+    class LegacyAdjuster
```

Comment:
If this is the only solution provided in core, I don't think we should call it "Legacy". Or am I missing something?

Reply:
The plan is another adjuster that does not call `Spree::Adjustment#recalculate`, as that code path is very difficult to optimize and quite error-prone in some situations. This one does call that method, which I think should be deprecated once we have an alternative.

Reply:
The adjustment should not be recalculated using its source; I'd much rather load all promotions and determine their eligibilty first, and then add, remove or change adjustments. The way it is done right now falls apart when using multi-level (orde,r line item, shipment, shipping rate) rules.

Reply:
I agree it's not the best way to handle the recalculation, but we risk that this legacy class ends up in a new version without the new option. My proposal is:

- rename it into something that works no matter if there's the new option available. Some semantic name that describes what it does would be awesome (and I know, hard to find as hell 😄 ). 
- merge it
- if we merge the new option before the next minor/major, we are still allowed to rename this class without deprecating it, because it has never been released
- if we don't merge the new option before the next minor/major, we can still deprecate the class, even if doesn't have the legacy suffix in its name. 

What do you think?